### PR TITLE
Release/session

### PR DESCRIPTION
**Pull request by geeknam**
- Add examples
- Use requests sessions
- Add more logging

### CODE REVIEW

File: gcm/gcm.py [hunk 1]
```diff
@@ -133,29 +138,29 @@ def validate(self, options):
...
 
 class PlaintextPayload(Payload):
-
     @property
     def body(self):
         # Safeguard for backwards compatibility
         if 'registration_id' not in self.__dict__:
             self.__dict__['registration_id'] = self.__dict__.pop(
-                'registration_ids', None
+                    'registration_ids', None
```
Comment (geeknam) [inline gcm/gcm.py]: What's with these indents? @mnemonicflow

File: gcm/gcm.py [hunk 2]
```diff
@@ -284,10 +298,18 @@ def make_request(self, data, is_json=True):
...
-            self.url, data=data, headers=headers,
-            proxies=self.proxy, timeout=self.timeout,
-        )
+        new_session = None
+        if not session:
+            session = new_session = requests.Session()
+
+        try:
+            response = session.post(
+                    self.url, data=data, headers=headers,
```
Comment (geeknam) [inline gcm/gcm.py]: What's with these indents? @mnemonicflow

File: gcm/gcm.py [hunk 3]
```diff
@@ -307,30 +329,31 @@ def make_request(self, data, is_json=True):
         # Failures
         if response.status_code == 400:
             raise GCMMalformedJsonException(
-                "The request could not be parsed as JSON")
+                    "The request could not be parsed as JSON")
```
Comment (geeknam) [inline gcm/gcm.py]: What's with these indents? @mnemonicflow

File: gcm/gcm.py [hunk 4]
```diff
@@ -307,30 +329,31 @@ def make_request(self, data, is_json=True):
...
+    @staticmethod
+    def raise_error(error):
         if error == 'InvalidRegistration':
             raise GCMInvalidRegistrationException("Registration ID is invalid")
         elif error == 'Unavailable':
             # Plain-text requests will never return Unavailable as the error code.
             # http://developer.android.com/guide/google/gcm/gcm.html#error_codes
             raise GCMUnavailableException(
-                "Server unavailable. Resent the message")
+                    "Server unavailable. Resent the message")
```
Comment (geeknam) [inline gcm/gcm.py]: What's with these indents? @mnemonicflow

**Comment by geeknam** [comment]
Will get the release out but someone needs to fix the indents @mnemonicflow

**Comment by mnemonicflow** [comment]
@geeknam As per https://www.python.org/dev/peps/pep-0008/#code-lay-out "continuation lines with more indentation included to distinguish this from the rest" is totally alright and this is a reformatting rule that I've run the code through inside PyCharm, but I'm not sure this project adheres to PEP8 coding style standards.
After a second look there are other cases where more spaces on continuation lines can be added by the above rule. If you want I can undo this or I can reformat again to apply the rule across the whole code.

**Pull request merged**